runtime/lib/ttnn/runtime.cpp

```diff
@@ -56,22 +56,40 @@ tt::target::DataType getTensorDataType(Tensor tensor) {
   return utils::fromTTNNDataType(nnTensor.dtype());
 }
 
-Device openDevice(std::vector<int> const &deviceIds,
-                  std::vector<std::uint8_t> const &numHWCQs) {
-  assert(deviceIds.size() == 1 && "Only one device is supported for now");
-  assert(numHWCQs.empty() && "HWCQs are not supported for now");
-  auto &device = ::ttnn::open_device(deviceIds.front(), kL1SmallSize);
-  device.enable_async(true);
-  return Device::borrow(device, DeviceRuntime::TTNN);
+size_t getNumAvailableDevices() {
+  return ::tt::tt_metal::GetNumAvailableDevices();
+}
+
+Device openDevice(DeviceIds const &deviceIds, size_t numHWCQs) {
+  size_t numDevices = ::tt::tt_metal::GetNumAvailableDevices();
+  assert(numDevices == deviceIds.size() &&
+         "Currently only support opening all available devices");
```

Comment:
@jnie-TT what is causing this constraint? It would be ideal to avoid it because, this way, the public API is broken. If not avoidable, let's sync so that FE side is prepared...

Reply:
Hey @pilkicTT so the idea here is that the user opens up all devices up front, then the compiler generates get_device ops that create subviews of the entire device mesh. This way if we have multiple programs running on different grid sizes, we don't need to dynamically check/re-open sub-meshes of devices. If this clashes with plans on FE we should definitely sync up, does FE have cases where only a subset of all devices will be opened? How does FE determine which devices should be opened?

Reply:
On the FE side, we are currently opening all devices one by one (before starting to execute a binary), storing their handles and closing them once we finish execution. As we are running everything on a single chip, it didn't really matter at the time and this was the simplest way... However, now when we are adding multi-chip support, we can change this if needed...

What I don't get from your description, is why does the function `openDevice()` have `deviceIds` when all of them will be opened? If we'll open each device up front, then it doesn't make sense to me to have a parameter for selecting devices? I was guessing that that is a temp workaround, but from your description it doesn't look that way... I am probably missing something... 😄

Reply:
We still have device IDs for 2 reasons: 1. We want to keep the user-facing API consistent. 2. In the future, we also want to potentially support opening sub meshes (not all devices) - for example, the user could run two independent flatbuffers on two sets of mesh devices, but for simplicity for now in the beginning of multi-chip bringup, we want to assert that all devices are opened